### DISCUSSION TITLE
`post_logout_redirect` without `id_token_hint`

### DISCUSSION THREAD

**Discussion by davidthor** [ory/hydra#2544] [Q&A] [answered]
I'm currently trying to implement the logout flow with Hydra, but can't seem to get the flow to redirect back to a specified URL. The URL I'm specifying is whitelisted for the client in `post_logout_redirect_uris`, but I don't have access to the `id_token` to send with the logout request. This is unfortunately a limitation of the library I'm using to implement the oauth flow, `@nuxtjs/auth`, whereby it doesn't persist the id_token value for me to include.

Is there any way for me to hint at the associated client w/out passing the `id_token_hint` field so I can use a specified `post_logout_redirect_uri` instead of the default for the Hydra instance?

**Comment by aeneasr** [comment, marked as answer]
Unfortunately not because we can not identify the client which initiated the logout request. I believe this is documented here: https://www.ory.sh/hydra/docs/concepts/logout/

**Reply by davidthor** [reply]
I saw the documentation, but I was curious as to why I can't just specify the client ID directly instead of having to send it as part of the id_token. Would there be any security or usability reason not to allow that as part of the flow?

**Reply by aeneasr** [reply]
Yes, that has security implications. Not validating the client ID opens up vulnerabilities: open redirect (register a new client with a random redirect URL), denial of service (delete session for a client you don't own) and probably a few others as well!

**Reply by davidthor** [reply]
Sorry for not being clear, but I wasn't suggesting we forgo the client ID, but rather that we specify the client ID outright instead of requiring it be bundled inside the id_token. Something like `/oauth2/sessions/logout?client_id_hint=...` instead of `/oauth2/sessions/logout?id_token_hint=...`.

**Reply by aeneasr** [reply]
Validation as in verifying a cryptographic signature (ID token) that originates from one of Ory Hydra's private keys.

**Reply by davidthor** [reply]
But what’s the attack vector that verifying the signature protects against? The return url still has to be whitelisted by the matching client and it seems like checking the ID token signature doesn’t have any added benefit in this case.

**Reply by aeneasr** [reply]
```
curl /oauth2/sessions/logout?client_id_hint=1234
curl /oauth2/sessions/logout?client_id_hint=1235
curl /oauth2/sessions/logout?client_id_hint=1236
curl /oauth2/sessions/logout?client_id_hint=1237
curl /oauth2/sessions/logout?client_id_hint=1238
```

without `client_id_hint=1234` actually requesting the logout but instead it's an attacker.

**Reply by davidthor** [reply]
Doesn’t it still require the attacker to have a cookie representing the user for logout to do anything at all? I can still go through a logout flow without the id_token_hint, I just can’t specify a return_to address. Isn’t it spoofing there return_to that this is protecting against?

**Reply by aeneasr** [reply]
No the cookie is owned by the user’s browser.

With id_token_hint we do not process front/backchannel logout so it’s not a DoS attack surface.

**Reply by davidthor** [reply]
Oh, I didn't realize that front/back channel logout wasn't performed in that flow. If the cookie/session isn't important as part of this logout flow, can it just be an ajax POST request instead of a browser redirect?

**Reply by aeneasr** [reply]
I think we got lost somewhere :/ The cookie is important for the logout, otherwise we can't log the user out? As of now, I can't think of something that would allow logout with a flexible return URL that doesn't required verification of the client by checking something that can be authenticated (like an OpenID Connect ID Token). I fear that there's not really a way around that.

**Reply by davidthor** [reply]
> ```
> curl /oauth2/sessions/logout?client_id_hint=1234
> curl /oauth2/sessions/logout?client_id_hint=1235
> curl /oauth2/sessions/logout?client_id_hint=1236
> curl /oauth2/sessions/logout?client_id_hint=1237
> curl /oauth2/sessions/logout?client_id_hint=1238
> ```
> 
> without `client_id_hint=1234` actually requesting the logout but instead it's an attacker.

I think this is what confused me. I understood that an attacker could try to brute-force client_id_hint values, but I didn't understand what that would allow them to do aside from potential DDoS the endpoint since they would have to have the users session cookie in their browser for the request to do anything. 

If the requester has both the session cookie, the client_id associated with it, and a whitelisted return_to URL, why shouldn't they be able to make that request? What extra benefit does validating the id_token_hint provide over simply validating the session cookie and that the client_id is associated with the return_to URL?

**Reply by aeneasr** [reply]
Accepting any client ID is a https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html

**Reply by davidthor** [reply]
But isn't that what the whitelisted return_to values in the config are for? It's not like someone could find a valid client_id and then just redirect the user wherever they want.

**Reply by aeneasr** [reply]
>  It's not like someone could find a valid client_id and then just redirect the user wherever they want.

That's exactly what the problem is. Ory Hydra supports Dynamic Client Registration and most platforms offer a "developer account" (Google, GitHub, ...) where anyone can create an OAuth2 Client. This is the OAuth2 use case.

Just because, in your environment, you have tight control over this, doesn't mean that it applies to others. Therefore, Malice can trick Alice into opening a logout URL which points to a phishing login page, for example. The link looks valid ("it's from ory.sh!") so the redirect has to be valid too, right?

Sorry, but we're going in circles. Please accept that there are security concerns, and that we have spent a lot of time on designing this. There are so many implications when you "just skip validation". I wish OAuth2 wasn't such a PITA all of the time, but that's the reality of this protocol family. If we could make it easier, I would!

**Reply by davidthor** [reply]
> That's exactly what the problem is. Ory Hydra supports Dynamic Client Registration and most platforms offer a "developer account" (Google, GitHub, ...) where anyone can create an OAuth2 Client. This is the OAuth2 use case.

That's what I was missing – totally forgot about that use-case. Thanks for pointing it out!

> Sorry, but we're going in circles. Please accept that there are security concerns, and that we have spent a lot of time on designing this. There are so many implications when you "just skip validation". I wish OAuth2 wasn't such a PITA all of the time, but that's the reality of this protocol family. If we could make it easier, I would!

No problem at all. I opened it as a discussion instead of an issue because I was confident there _was_ an answer to the question and that someone on your team had likely thought of it. Thanks for your patience!

**Reply by Csirme** [reply]
I opened the same discussion，why not use the "client_id" + "redirect_uri" to verification, then client can decide redirect to . @aeneasr
